Numpy-style broadcasting for all mathematical functions

gchanan opened the pull request:
all functions mentioned in: implements all functions mentioned in https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes

gchanan committed:
Expand improvements

1) Rename calculateExpandGeometry to inferExpandGeometry for consistency
2) Simplify inferExpandGeometry implementation by using a single pass
   through dimensions
3) Implement a two operand expansion, expand2.
4) Implement versions that return error code to use for fallback to
equal nElem support.

gchanan committed:
Testing over the following:
1) CPU tensor out-of-place functions
2) CPU tensor in-place functions
3) GPU tensor out-of-place functions
4) GPU tensor in-place functions
5) torch. functions
6) Fallback semantics (use pointwise nElem matching rather than broadcasting)

gchanan committed:
Add documentation including links to numpy broadcasting semantics.

gchanan committed:
Add Broadcast plugin.

gchanan committed:
pow, fmod, remainder also should fallback.

This behavior isn't listed in the docs, but the tests depend on it.

gchanan committed:
Test fixes for keepdim=False, suppress warnings on backwards-compatible behavior.

gchanan committed:
Backwards compatible Spatial Normalizations / CrossMapLRN.

gchanan committed:
Implement/test broadcasting semantics for comparison ops.

gchanan committed:
Add Infer Size N, for expansion of fused operations.

gchanan committed:
Change order of Broadcast specification.

Since fused ops require broadcasting self over multiple other arguments,
it is simpler to specify broadcast on self rather than the other
way around.

gchanan committed:
Breakup in place broadcast to better handle multiple arguments.

gchanan committed:
Support "fused" ops: addcmul/addcdiv.

gchanan committed:
Docs for addcdiv, addcmul

gchanan committed:
Proof that broadcasting 3 args (expand3) is equivalent to
breaking up operation.

gchanan committed:
Revert "Change keepdim default to False."

This reverts commit e124790cb2b6675a4b6edf64620a7eb7f7228b29.

Note the original commit message is incorrect; this changes keepdim
back to false.

gchanan committed:
Revert "Restore examples with keepdim=True default."

This reverts commit 6fab62173e842bbf550de1c68cfae507ca35b800.

gchanan committed:
Add optional warning for backwards incompatible broadcast.

Setting torch.utils.backcompat.broadcast.warning.enabled=True
will cause Python warnings in the case where broadcast occurs
but previously 1-d view style pointwise ops occured.

gchanan committed:
Document Broadcast Plugin.

gchanan commented:
Here is numpy-style broadcasting for pointwise math and reduction functions.

This also changes the keepdim default from True to False.

I have a longer writeup in the works categorizing the numpy semantics and showing that these are the only backwards incompatible changes necessary to get numpy-style broadcasting (i.e. what's left from here only adds functionality that would currently give you an error).

I'll link the writeup when that's done, but feel free to review before then (although you may wish to hold off on merging).

soumith commented:
let's aim to merge this after the nips deadline on Friday

ezyang commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,76 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
-TH_API void THLongStorage_calculateExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est) {
-  ptrdiff_t ndim = THLongStorage_size(sizes);
-  long numUnsqueezed = ndim - tensorDim;
+TH_API int THLongStorage_inferSize2(THLongStorage *output, long *sizesA, long dimsA, long *sizesB, long dimsB, int raiseErrors) {
```
Given the sizes and dimensions of two tensors, write out the final size when the two tensors are broadcasted together.

ezyang commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,76 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
-    expandedSizes[i] = 1;
-    expandedStrides[i] = expandedSizes[i+1] * expandedStrides[i+1];
-  }
+TH_API int THLongStorage_inferExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est, int raiseErrors) {
```
Given an existing tensor with size tensorSizes, strides tensorStrides and dimension tensorDim, compute the new expanded size (saved to esz) and new necessary stride (saved to est) necessary to broadcast (expand) this tensor to sizes.

ezyang commented on torch/lib/TH/generic/THTensor.c:
```diff
@@ -285,21 +285,80 @@ void THTensor_(resize5d)(THTensor *self, long size0, long size1, long size2, lon
+  return 0;
+}
+
+int THTensor_(expand2)(THTensor *ra, THTensor *rb, THTensor *opa, THTensor *opb, int raiseErrors) {
```
Given two tensors, expand them so that they have identical dimensions and sizes, following broadcasting semantics.

ezyang commented on tools/cwrap/plugins/Broadcast.py:
```diff
@@ -0,0 +1,106 @@
+from string import Template
+
+
+class Broadcast(CWrapPlugin):
```
Adds boilerplate to broadcast arguments to be identical sizes, before running the algorithm proper. Only works for binary operators.

ezyang commented on test/test_torch.py:
```diff
@@ -939,6 +939,156 @@ def test_arange(self):
+            self.assertEqual(t0.size(), r0.size())
+            self.assertEqual(t1.size(), r1.size())
+
+            # case 4: not broadcastable and not nEleme equal -- tested by test_fallback
```
s/nEleme/nElem/

ezyang commented on torch/lib/TH/THStorage.c:
```diff
@@ -104,6 +104,49 @@ TH_API int THLongStorage_inferSize2(THLongStorage *output, long *sizesA, long di
   return 0;
 }
 
+TH_API int THLongStorage_inferSizeN(THLongStorage *output, int n, long **sizes, long *dims, int raiseErrors) {
```
Given n tensors with given sizes and dims, broadcast all of them to a single size, which is recorded in output.

ezyang reviewed (commented):
I love it. I went through and wrote some documentation for the new functions while reading. I didn't do a very careful "is this logic correct" code review.

gchanan commented:
@ezyang thank you, I will incorporate these suggestions.

If you are interested in how this works / future plans, you may find these notes interesting: https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes.

gchanan committed:
Include simple broadcasting example and demonstrate lining up trailing dimensions.

albanD commented:
@gchanan could it be possible to have a global switch (at compile time if not possible at runtime) that would disable the automatic broadcasting? For example that just disable the cwrap plugin?
I (and maybe other people) prefer to use view/expand and not have the broadcast that can hide subtle bugs.

gchanan commented:
@albanD good question and I understand the motivation for tightly controlling your own code.  What I'm not sure about is how to make the distinction between your code and library code (or if that distinction is even well defined).

To support what you want as written, we'd have to ensure that our libraries work both with and without broadcasting, which will greatly increase the maintenance burden.  It's also not clear to me the distinction between user and library code is really clear anyway: let's say some NN layer is written in such a way that it e.g. multiples the input tensor by a (4,1) tensor and you pass in a (4) tensor.  Was it library or user code that caused broadcasting?

What I have now is a UserWarning if your code does not broadcast, but uses the old (deprecated) 1-d pointwise operations (you previously only needed nElem to match up for many functions).  I could also add an optional warning for the backwards-incompatible case, that is, the sizes don't match, so you would have used the 1-d pointwise operations, but you are now broadcasting.

It sounds like what you want is a warning if broadcasting changes the sizes at all (I'd be hestitant to make it an error, not a warning, given the argument about library code above).  That shouldn't be too difficult to implement, although I'm not sure how useful it will be if our library code broadcasts a lot (you might get a lot of warnings).  What do you think?

albanD commented:
Ho right I assumed that our library will stay broadcasting free (as it is right now).
I agree that keeping it that way may be a bit more complex for people that want to send new PR and always use broadcasting. But here the change are quite minor as you would just need to slightly adapt new PR and the existing code will not change and work as is.
I guess @soumith and @apaszke will need to decide if the possibility of disabling the broadcasting is worth this maintaining cost ?

pavanky commented:
What about a per tensor switch instead of a global switch ? This will however increase the complexity in user code.

gchanan commented:
If you pass in a non-broadcasting tensor to the library, does it switch to a broadcasting tensor?  If it doesn't, you have the same issue of the library having to work with both (and having to define in what cases the flag is copied).

gchanan committed:
Add broadcast support for fused-matmul broadcasting.
Functions are: addmm, addbmm, addr, addmv, baddbmm.

gchanan committed:
Add broadcasting support for matmul.

gchanan committed:
Add optional warning for backwards incompatible keepdim.
Setting torch.utils.backcompat.keepdim.warning.enabled=True
will cause Python warnings in the case where the default value
of keepdim is used for 1-d reductions.

Also specify keepdim via kwargs in library so these warnings have less
noise.

gchanan committed:
Add backwards incompatibility docs.

gchanan committed:
Improve in-place broadcasting back compat warning message and fix an issue
where the deprecated warning would not be printed.

gchanan committed:
Properly clean up expand error cases.

gchanan committed:
Use THSize_isSameSizeAs, instead of THTensor_(isSameSizeAs) in order
to compare sizes of tensors with different data types.

gchanan committed:
Add broadcasting support for masked_copy, masked_fill.

gchanan committed:
Add broadcasting support for map_, map2_.

gchanan committed:
Add broadcasting support for copy_, simplify code generation by
moving a lot of currently generated code to expand_utils.

gchanan committed:
Add documentation for copy_.

gchanan committed:
explicit Ptr constructors, fast transposed copy.

gchanan committed:
Fix test_broadcast for cuda tensors, since map_, map2_ not implemented.

gchanan committed:
Add a broadcast parameter to copy_, use it in the library in
cases where there is non-broadcasting calls exposed by the tests.

gchanan committed:
Ensure warnings are repeated in test, necessary in python2.

gchanan commented:
I think this is ready for review and now implements all functions mentioned in https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes (i.e. it no longer includes only pointwise mathematical and comparison functions like the title says).

Re: turning off broadcasting, the only case I implemented this for was for copy, which allows you to pass a "broadcast" parameter (default: True).  There were a number of places in the code where it was clearly intended to copy tensors as 1-d, which wasn't really true for other functions.

This PR now also includes some warnings that you can enable to detect backwards incompatible changes.  In particular:

1) Setting torch.utils.backcompat.broadcast.warning.enabled=True will cause Python warnings in the case where broadcast occurs but previously 1-d view style pointwise ops occured.

2) Setting torch.utils.backcompat.keepdim.warning.enabled=True will cause Python warnings in the case where the default value of keepdim is used for 1-d reductions.

I've manually enabled these and verified that the only places these warnings are triggered are in the tests are directly in tests, i.e. there are no library calls (at least that are covered by tests) where we rely on the old behavior.  This should make turning on these warnings less noisy.

gchanan committed:
Fix lint errors.

gchanan commented:
rebased the commits.

killeent commented on docs/source/notes/broadcasting.rst:
```diff
@@ -0,0 +1,108 @@
+Many PyTorch operations support :any:`NumPy Broadcasting Semantics <numpy.doc.broadcasting>`.
+
+In short, if a PyTorch operation supports broadcast, then its Tensor arguments can be
+automatically expanded to be of equal sizes (without making copies of the data).
```
Nit: once and for all I would like us to settle on either "size" or "shape"

gchanan replied:
Agreed, and I wasn't consistent in this PR.  In general I prefer "shape" but there is already a size() function and a torch.Size object.  I generally used size/sizes except in cases I thought shape was significantly clearer.

killeent commented on tools/cwrap/plugins/Broadcast.py:
```diff
@@ -0,0 +1,294 @@
+#                              arguments to broadcast specified argument (usually "self") against
+# [inplace] will generate code for in-place function, which doesn't allow the in-place
+#           argument to be broadcast
+# [fallback] if tensors aren't broadcastable, preserves "element number" pointwise behavior,
```
Nit: make this `broadcast_fallback` or something that makes it explicit what behavior we are "falling back" from.

killeent replied:
Yes, I realized this later on.

killeent commented on tools/cwrap/plugins/Broadcast.py:
```diff
@@ -0,0 +1,294 @@
+#           argument to be broadcast
+# [fallback] if tensors aren't broadcastable, preserves "element number" pointwise behavior,
+#            where only number of elements need to match, and tensors are viewed as 1-dimensional.
+# [dims] if the tensors shouldn't be broadcast to specific tensor or tensors, but a combination
```
It's a little clear what `dims` is used for, can you expand upon this.

killeent commented on torch/csrc/Module.cpp:
```diff
@@ -466,6 +466,54 @@ PyObject *THPModule_addDocStr(PyObject *_unused, PyObject *args)
 
+
+
+PyObject *THPModule_inferSize(PyObject *_unused, PyObject *args)
```
Why is this added here? What is it used for?

killeent commented on torch/csrc/copy_utils.h:
```diff
@@ -50,24 +51,88 @@ inline PyObject * THPCopyMethod(const THPCopyList& v, PyObject *self, PyObject *
+  PyObject *async = Py_False;
+  PyObject *broadcast = Py_True;
+  static char *kwlist[] = {"source", "async", "broadcast", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|OO:copy_", kwlist, &src, &async, &broadcast)) {
```
Is this not able to verify the types of keyword arguments? I can't remember.

gchanan replied:
sure and I'll change the StorageCopy too.

killeent commented on torch/csrc/copy_utils.h:
```diff
@@ -50,24 +51,88 @@ inline PyObject * THPCopyMethod(const THPCopyList& v, PyObject *self, PyObject *
     TensorDst* dst = THPTypeInfo<TensorDst>::cdata(dst_);
     TensorSrc* src = THPTypeInfo<TensorSrc>::cdata(src_);
 
+    TensorSrc *src_save = src;
```
Can you explain what is going on here with broadcast?

killeent commented on torch/csrc/cuda/expand_utils.cpp:
```diff
@@ -0,0 +1,188 @@
+#include "torch/csrc/cuda/THCP.h"
```
Why do we need these files (e.g. the same for csrc below)? Why can't we just call `THTensor_(new)`?

gchanan replied:
This is tricky, and I'll add a comment, although it's already done in a number of places in the code.  The functions in expand_utils.h/expand_utils-inl.h are defined in terms of the LIBRARY_STATE and related macros (because the copy functions are defined the same way).  So, there are actually two versions of each function (with and without THCState) and we avoid redefining the non-cuda ones by including THCP.h (that in terms defines the macros) and then including those files.

apaszke replied:
Can we maybe create a generic dir in utils and generate these specializations via macros? It's going to be much less repetetive

killeent commented on torch/csrc/expand_utils.h:
```diff
@@ -0,0 +1,155 @@
+#ifndef THP_EXPAND_UTILS_H
```
Can you document this file? It's very unclear what is going on.

gchanan replied:
Same as above.

killeent commented on torch/csrc/utils.cpp:
```diff
@@ -551,3 +551,33 @@ void THPPointer<THPGenerator>::free() {
 
 template class THPPointer<THPGenerator>;
+
+static bool backCompatBroadcastWarn = false;
```
Someone better at C++ than I can evaluate whether or not this works as a global variable?

gchanan replied:
it's not global, it's static :).

killeent commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
+  THLongStorage_resize(output, ndim);
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
```
Above in `THSize_isSameSizeAs` we return 1 on success, 0 on failure. Here we return 0 on success, -1 on failure? It would be nice for these to be consistent?

gchanan replied:
I don't think they should be consistent -- expand is an error code / success indicator, THSize_isSameSizeAs should return "true" if those are the same size, there's no error code / success in there.

For clarity it's probably just better to define an error code type THError_t.

killeent reviewed (commented):
In the future I think it would be helpful if we could come up with some sort of mechanism for breaking up PRs into smaller chunks, although I'm not blaming you for that @gchanan.

Due to the large scope of this PR, I didn't really look at the logic so much, I guess I'll just have to trust you. I added a few nits and questions where I didn't understand generally why things were done.

gchanan committed:
Clarify a number of comments.

apaszke commented on torch/csrc/expand_utils.h:
```diff
@@ -0,0 +1,155 @@
+      std::ostringstream warn;
+      warn << tensor_name << " and " << to_expand_name << " do not have the same shape, but are "
+           << "broadcastable, and have the same number of elements.  Changing behavior in a backwards incompatible "
+           << "manner to broadcasting rather than viewing as 1-dimensional.";
```
Shouldn't we warn even if they are not broadcastable? I thought we're going to deprecate that old behaviour.

apaszke commented on torch/lib/TH/THSize.h:
```diff
@@ -0,0 +1,13 @@
+#include <stddef.h>
+
+// THTensor functions that would work on a THSize if we had such a class in C++,
+// i.e. THTensor functions that depend only on the shape of the tensor, not the type.
```
THTensor functions that depend on the shape, but not type are all `THLongStorage` functions, right? That's what we've used for `inferSize` etc. I think it's a better place for it

apaszke replied:
I think it's better to be consistent. Right now, we use LongStorage for such purposes, and I'd rather keep it like that, with an option to change it as you propose in a later PR.

colesbury replied:
I don't think we should be adding additional uses of THLongStorage to represent sizes. It was a mistake to represent sizes that way that only sort-of made sense in the context of Lua.

We already use `long*` pointers in `resizeNd` and `setStorageNd`.

It's also bad because we take the THLongStorage.size -> THTensor.nDimension, but one is an `int` and the other is a `long`. This requires either `-fpermissive` everywhere or extra casts.

apaszke commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
+        return -1;
+      }
+    }
+    expandedSizes[ i ] = sizeA;
```
wouldn't it be simpler to write it as
```c
if (sizeA == sizeB || sizeA == 1 || sizeB == 1) {
  expandedSizes[i] = THMax(sizeA, sizeB);
} else {
  // Error
}
```

apaszke commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
+      }
+    }
+
+    expandedSizes[ i ] = max_dim_size;
```
as above

apaszke commented on torch/lib/TH/THStorage.c:
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
+  return 0;
+}
+
+TH_API int THLongStorage_inferExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est, int raiseErrors) {
```
what does this function do? what are esz and est? Expected sizes and expected strides?

gchanan replied:
This is just calculateExpandGeometry, I just renamed it to be consistent with other uses of "infer" and simplified the implementation.

To answer your question, I think esz is expandedSizes and est is expandedStrides.  I'll just rename this in this PR, I agree it's confusing.

apaszke commented on torch/lib/TH/generic/THTensor.c:
```diff
@@ -285,23 +285,103 @@ void THTensor_(resize5d)(THTensor *self, long size0, long size1, long size2, lon
-  THArgCheck(THLongStorage_size(sizes) >= THTensor_(nDimension)(tensor), 1, "the number of sizes provided \
-      must be greater or equal to the number of dimensions in the tensor");
-  THArgCheck(THTensor_(nDimension)(tensor) > 0, 0, "can't expand an empty tensor");
+THTensor* THTensor_(newExpand)(THTensor *tensor, THLongStorage *sizes, int raiseErrors) {
```
Why is this new flag necessary? It's not at all obvious why expand functions should have a flag for errors? I think it's better to try to avoid changing the base of tensor API

gchanan replied:
I think I can get rid of it.  I need the raiseErrors flag for other expand functions to support fallback and at one point the Broadcast code called this function, but it doesn't anymore.  So should be safe to remove it here but keep it for expand, expand2, expand3.

apaszke replied:
Can't you just use a try-catch in THP to handle the cases when it fails?

gchanan replied:
is it guaranteed that THError (what gets called now if raiseErrors = 0) will throw an exception?  The function is settable, right?  So could e.g. exit.

I guess we could move all of this to THPP and add real exception types?

apaszke replied:
Yes, it is guaranteed in PyTorch. We depend on it all over the place (see how `HANDLE_TH_ERRORS` macro is defined in `torch/csrc/Exceptions.h`). I don't think having this flag as part of the interface is a good idea. It's only for this single use case.

apaszke commented on torch/tensor.py:
```diff
@@ -300,7 +302,24 @@ def __matmul__(self, other):
+
+            # reshape batches back into result
+            total_expansion = expand_batch_portion + (self_exp_size[-2], other_exp_size[-1])
+            return self_expanded.bmm(other_expanded).view(*(total_expansion))
```
doesn't `bmm` expand its arguments?

apaszke replied:
So does numpy broadcast inputs to matrix multiplication at all?

gchanan replied:
Not for conventional 2-d matrix multiplication, e.g. (7,5) @ (1,3) won't broadcast to (7,5) @ (5,3).

fmassa replied:
Given that `mm`, `bmm`, `ger` are low-level functions that do not perform broadcasting, what about explicitly stating that in the docs, and add a `torch.matmul` function equivalent to `__matmul__`, with proper documentation? For the moment, the `@` operator is not documented, and only potentially used by py3 users.

apaszke replied:
Yeah sounds like a nice thing to have

gchanan replied:
Agreed, this is a good idea.  I'm going to open a separate issue for it given this issue is large enough and ready to merge already.

gchanan replied:
Filed Issue https://github.com/pytorch/pytorch/issues/1756 for this.

fmassa replied:
We probably need to fix the `Variable` case for `__matmul__` to reflect the behavior of the new `__matmul__` as well.

apaszke commented on torch/utils/backcompat/broadcast/warning/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+    enabled = property(get_enabled, set_enabled)
+
+sys.modules[__name__] = Warning()
```
1. Don't use old style classes (inherit from `object` explicitly)
2. Wouldn't it be simpler to have a `warning` variable in `torch.utils.backcompat.broadcast` instead of doing this hack?

gchanan replied:
On 2. I think I did this because properties don't work outside of objects, see https://stackoverflow.com/questions/17007886/property-method-without-class.  At least some quick attempts don't seem to work.

apaszke replied:
I meant that you could do this instead of hacking `sys.modules`:
```python
# torch/utils/backcompat/broadcast.py
class Warning(object):
    ... # declare the property here
warning = Warning()
```
Also, maybe let's just have `torch.utils.backcompat.broadcast_warning` and `torch.utils.backcompat.keepdim_warning`? You could reuse the same class for this (just give the getter/setter C function to the constructor of both instances).

gchanan commented on docs/source/notes/broadcasting.rst:
```diff
@@ -0,0 +1,112 @@
+    
+    # but:
+    >>> x=torch.FloatTensor(5,2,4,1)
+    >>> y=torch.FloatTensor(3,1,1)
```
sure.

gchanan commented on test/common_nn.py:
```diff
@@ -103,7 +103,7 @@
         module_name='LogSoftmax',
         input_size=(1, 3, 10, 20),
-        reference_fn=lambda i, _: torch.exp(i).div_(torch.exp(i).sum(1).expand_as(i)).log_(),
+        reference_fn=lambda i, _: torch.exp(i).div_(torch.exp(i).sum(1, False).expand_as(i)).log_(),
```
Nope.

gchanan commented on torch/_tensor_str.py:
```diff
@@ -67,7 +67,7 @@ def set_printoptions(
 def _number_format(tensor, min_sz=-1):
     min_sz = max(min_sz, 2)
-    tensor = torch.DoubleTensor(tensor.nelement()).copy_(tensor).abs_()
+    tensor = torch.DoubleTensor(tensor.size()).copy_(tensor).abs_().resize_(tensor.nelement())
```
sure.

gchanan commented on torch/csrc/expand_utils-inl.h:
```diff
@@ -0,0 +1,175 @@
+}
+
+template <>
+THLongTensor *newForExpand(LIBRARY_STATE_TYPE_NOARGS) {
```
We expand on the CPU because the CPU copy methods have THCState if the source is on the GPU.

gchanan commented on test/test_torch.py:
```diff
@@ -939,6 +944,385 @@ def test_arange(self):
+                small2 = cast(small2)
+                small2Expanded = small2.expand(*dims_full)
+
+            if fn not in fns_no_out_place:
```
yes, good suggestion.   I was going to see if I could get rid of the exceptions (e.g. just add a max_), but the issue become large enough by itself.

gchanan commented on test/test_torch.py:
```diff
@@ -939,6 +944,385 @@ def test_arange(self):
+                    if fn == "lerp":
+                        return t0_fn(t1, 0.5)
+                    elif fn == "masked_copy":
+                        return t0_fn(t1 < 0.5, t1.expand_as(t0))
```
No, because it's not really masked_copy, more like masked scatter and the shape of the source has nothing really to do with the shape of the underlying tensor, so we don't broadcast it.

It's just unnecessarily confusing for me to have written the test to use the expanded/non-expanded tensors, though; I'll just use a 1-d arange of the proper length or something.

gchanan committed:
Incorporate review comments:

1) Line up trailing dimensions in broadcast docs.
2) remove unnecessary expand_as in common_nn test.
3) use view in tensor_str instead of resize_.
4) newExpand remove raiseErrors change.
5) clarify expandedSizes/expandedStrides parameters in inferExpandGeometry.
6) simplify inferSize2/inferSizeN implementations.
7) use new-style classes for warning.

gchanan committed:
Move expand_utils-inl.h to generic/ and generate via macros.

gchanan committed:
Change async/broadcast copy arguments to be parsed as ints.

gchanan committed:
Rename check_fallback to check_backincompat_expand_warn for clarity.

gchanan committed:
Simplify python warning settings and cleanup tests.

gchanan commented:
@apaszke I incorporated all your suggestions minus the THLongStorage suggestion (due to the @colesbury comment above) and removing raiseErrors from the expand,expand2,expand3 functions, since that comment thread didn't conclude.

There were also new merge conflicts, so the last commit resolves them via a merge, but let me know if you want me to force push everything via a rebase.

apaszke commented:
I guess we can leave the THSize change, but it'd be better to remove the error flag.

gchanan committed:
Remove raiseErrors from THTensor functions, have THStorage functions
take an error_buffer to return a proper error message while being
able to handle memory management correctly from calling function.

gchanan commented:
ok, it took some effort to handle the error cases correctly without leaking memory, but I removed the error flag from THTensor.

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -485,7 +485,11 @@ PyObject *THPModule_inferSize(PyObject *_unused, PyObject *args)
+  char error_buffer[1024];
+  int ret = THLongStorage_inferSize2(sizes, size1->data, size1->size, size2->data, size2->size, error_buffer, 1024);
+  if (ret != 0) {
+    THError(error_buffer);
```
It's a python wrapper, you can just use `THPUtils_assert(ret != 0, error_buffer)`. We're not using `THError` anywhere here

gchanan committed:
Renamed masked_copy to masked_scatter in test, fix use of break/continue.

gchanan committed:
Add dist, atan2, lerp to fallback functions.

They weren't documented as having those semantics, but tests on
master show they do.

gchanan committed:
Clarify use of warn vs raise in expand_utils and don't catch exception
in Broadcast plugin when fallback = false.

gchanan committed:
Use THPUtils_assert rather than THError in torch/csrc/Module.

gchanan committed:
Fix Prod backward for broadcasting.

gchanan committed:
Rename autograd keepdim tests that now default to True.

gchanan committed:
Ensure warnings are repeated in python2 for tests.

soumith commented:
I fixed the remaining lint and merged this into master!

THIS WAS AWESOME @gchanan

soumith commented:
i forgot that parts of this PR have TH / THC etc. being touched upon. So I have to do the annoying reverse-merge scheme. So for now I reverted this PR on master (force-pushed, sorry), and I'll merge it in properly tomorrow-ish.

soumith commented:
this is now properly reverse-merged into master
